### PR TITLE
Reducing qcodes import times

### PR DESCRIPTION
**Pull request by edumur**
Hi all,

Following this discussion #4565 and this issue #4543 I am proposing this pull request to work on reducing qcodes loading time.

I have started with pandas.
I tried to propagate what @jenshnielsen did to make pandas import lazy-import.
I am not sure I have succeeded since I still see pandas being imported while doing `importtime-waterfall`.
I would gladly improve the request if someone help me on this.

If this becomes a successful endeavor, I will continue by making other import lazy-import.

### CODE REVIEW

**Comment by edumur** [comment]
And also, I do not no how to run qcodes test and check if I didn't break anything ^^'.
Again, if someone may help me.

**Comment by jenshnielsen** [comment]
Thanks, I will try to have a look asab and see if I can figure out why/if pandas is still being imported eagerly

**Comment by edumur** [comment]
Found out that pandas was imported by xarray so I passed xarray in lazy import.

On the data_set file, I got some error and had to write the type of some parameters as `'xr.Dataset'`, see line 693 for example. I do not know why and if this is correct.

**Comment by edumur** [comment]
Yep, forget what I wrote, it is fixed.

**Comment by jenshnielsen** [comment]
Looks good. Did you verify that you need to also make changes to the tests. Importing qcodes should not import the tests module. If it does that seems like a bug we should also fix.

**Comment by edumur** [comment]
You right, test file do not matters.
Could you remind me how to run the test locally?

**Comment by edumur** [comment]
Alright there is maybe the import of Monitor to work on but I am waiting opinion(s) from others.
On my pc it looks like it is 20% faster, can anyone look on their machine?

File: qcodes/dataset/plotting.py [hunk 1]
```diff
@@ -36,22 +37,13 @@
 log = logging.getLogger(__name__)
 DB = qc.config["core"]["db_location"]
 
-AxesTuple = Tuple[Axes, Colorbar]
-AxesTupleList = Tuple[List[Axes], List[Optional[Colorbar]]]
-# NamedData is the structure _get_data_from_ds returns and that plot_by_id
-# uses internally
-NamedData = List[List[DSPlotData]]
+if TYPE_CHECKING:
+    AxesTuple = Tuple[Axes, Colorbar]
```
Comment (jenshnielsen) [inline qcodes/dataset/plotting.py:41]: These typevars cannot be inside `if TYPE_CHECKING` but you can use "" around the imports

**Comment by edumur** [comment]
I should have include everything you said.
Please tell me if I missed anything.

File: qcodes/dataset/dond/do_nd_utils.py [hunk 2]
```diff
@@ -1,36 +1,49 @@
...
+    AxesTupleListWithDataSet = Tuple[
+        DataSetProtocol,
+        List["matplotlib.axes.Axes"],
+        List[Optional["matplotlib.colorbar.Colorbar"]],
+    ]
+    MultiAxesTupleListWithDataSet = Tuple[
+        Tuple[DataSetProtocol, ...],
+        Tuple[List["matplotlib.axes.Axes"], ...],
+        Tuple[List[Optional["matplotlib.colorbar.Colorbar"]], ...],
+    ]
```
Comment (jenshnielsen) [inline qcodes/dataset/dond/do_nd_utils.py:46]:
```suggestion
AxesTuple = Tuple["matplotlib.axes.Axes", "matplotlib.colorbar.Colorbar"]
AxesTupleList = Tuple[
    List["matplotlib.axes.Axes"], List[Optional["matplotlib.colorbar.Colorbar"]]
]
AxesTupleListWithDataSet = Tuple[
    DataSetProtocol,
    List["matplotlib.axes.Axes"],
    List[Optional["matplotlib.colorbar.Colorbar"]],
]
MultiAxesTupleListWithDataSet = Tuple[
    Tuple[DataSetProtocol, ...],
    Tuple[List["matplotlib.axes.Axes"], ...],
    Tuple[List[Optional["matplotlib.colorbar.Colorbar"]], ...],
]
```

As long as these are quoted I don't think they need to be wrapped in if typechecking and this breaks importing these types elsewhere

File: qcodes/data/data_set.py [hunk 3]
```diff
@@ -1,14 +1,17 @@
 """DataSet class and factory functions."""
+from __future__ import annotations
 
 import logging
 import time
 from collections import OrderedDict
 from copy import deepcopy
 from traceback import format_exc
-from typing import Any, Callable, Dict, List, Optional
+from typing import TYPE_CHECKING, Any, Callable, Dict, Optional
```
Comment (eendebakpt) [inline qcodes/data/data_set.py:9]:
```suggestion
from typing import TYPE_CHECKING, Any, Callable, Dict
```

File: qcodes/tests/dataset/helper_functions.py [hunk 4]
```diff
@@ -1,18 +1,24 @@
+from __future__ import annotations
+
 from functools import reduce
 from operator import mul
-from typing import Dict, List, Optional, Sequence, Tuple, Union
+from typing import TYPE_CHECKING, Dict, List, Optional, Sequence, Tuple, Union
```
Comment (eendebakpt) [inline qcodes/tests/dataset/helper_functions.py:5]:
```suggestion
from typing import TYPE_CHECKING, List, Optional, Sequence, Tuple, Union
```

File: qcodes/tests/dataset/helper_functions.py [hunk 5]
```diff
@@ -73,14 +81,18 @@ def verify_data_dict_for_single_param(datadict: Dict[str, np.ndarray],
...
-                                      names: Sequence[str],
-                                      shapes: Sequence[Tuple[int, ...]],
-                                      values):
+def verify_dataframe_for_single_param(
+    dataframe: pd.DataFrame,
+    names: Sequence[str],
+    shapes: Sequence[tuple[int, ...]],
+    values,
+):
+    import pandas as pd
```
Comment (eendebakpt) [inline qcodes/tests/dataset/helper_functions.py:90]:
```suggestion
    import pandas as pd  # pylint: disable=import-outside-toplevel
```
(and similar at other locations)

**Review by eendebakpt — COMMENTED** [review]
Thanks for picking this up! Can you fix the last linting errors? I will test import time on my machine

**Comment by edumur** [comment]
Sure, I was using `python -m mypy --ignore-missing-import -p qcodes` and getting:
```bash
qcodes\tests\dataset\test_export_info.py:9:2: error: Untyped decorator makes function "_make_basic_export_info" untyped  [misc]
qcodes\tests\parameter\test_combined_par.py:66: error: Unused "type: ignore" comment
qcodes\tests\parameter\test_combined_par.py:67: error: Unused "type: ignore" comment
qcodes\tests\parameter\test_combined_par.py:68: error: Unused "type: ignore" comment
qcodes\tests\legacy\test_combined_loop.py:35: error: Unused "type: ignore" comment
qcodes\tests\legacy\test_combined_loop.py:43: error: Unused "type: ignore" comment
qcodes\tests\legacy\test_combined_loop.py:51: error: Unused "type: ignore" comment
qcodes\tests\legacy\test_combined_loop.py:97: error: Unused "type: ignore" comment
qcodes\tests\legacy\test_combined_loop.py:105: error: Unused "type: ignore" comment
qcodes\tests\legacy\test_combined_loop.py:113: error: Unused "type: ignore" comment
qcodes\tests\legacy\test_combined_loop.py:162: error: Unused "type: ignore" comment
qcodes\tests\legacy\test_combined_loop.py:170: error: Unused "type: ignore" comment
qcodes\tests\legacy\test_combined_loop.py:178: error: Unused "type: ignore" comment
qcodes\tests\legacy\test_combined_loop.py:231: error: Unused "type: ignore" comment
qcodes\tests\legacy\test_combined_loop.py:239: error: Unused "type: ignore" comment
qcodes\tests\legacy\test_combined_loop.py:247: error: Unused "type: ignore" comment
qcodes\tests\dataset\test_dataset_loading.py:101:2: error: Untyped decorator makes function "test_get_run_attributes" untyped  [misc]
qcodes\tests\dataset\conftest.py:646:2: error: Untyped decorator makes function "_make_dummy_instrument" untyped  [misc]
qcodes\tests\drivers\test_lakeshore_325_legacy.py:11: error: Unused "type: ignore" comment
qcodes\tests\drivers\test_lakeshore_325.py:19: error: Unused "type: ignore" comment
Found 20 errors in 7 files (checked 609 source files)
```

So I thought it was fine.
Will get back with the changes.

File: qcodes/data/data_set.py (hunk 3 above)
Comment (edumur) [reply qcodes/data/data_set.py:9]: Looks like `from typing import TYPE_CHECKING, Sequence` would be enough

File: qcodes/tests/dataset/helper_functions.py (hunk 4 above)
Comment (edumur) [reply qcodes/tests/dataset/helper_functions.py:5]: Looks like `from typing import TYPE_CHECKING, Sequence` would be enough

**Comment by edumur** [comment]
Should be good

File: qcodes/__init__.py [hunk 6]
```diff
@@ -53,7 +53,6 @@
     VisaInstrument,
     find_or_create_instrument,
 )
-from qcodes.instrument_drivers.test import test_instrument, test_instruments
```
Comment (jenshnielsen) [inline qcodes/__init__.py:56]: This should be noted as a breaking change with documentation about where to import it from. (not that I think this is very commonly used but still)

File: qcodes/dataset/plotting.py [hunk 7]
```diff
@@ -9,15 +9,16 @@
...
-from matplotlib.colorbar import Colorbar
-from matplotlib.ticker import FuncFormatter
-from typing_extensions import Literal
+
+if TYPE_CHECKING:
+    import matplotlib
+    from matplotlib.axes import Axes
+    from matplotlib.colorbar import Colorbar
+    from typing_extensions import Literal
+    from matplotlib.ticker import FuncFormatter
```
Comment (jenshnielsen) [inline qcodes/dataset/plotting.py:21]: check if import of FuncFormatter can be complely removed

File: qcodes/dataset/plotting.py [hunk 8]
```diff
@@ -36,22 +37,14 @@
...
-# First find the kwargs that belong to subplots and than add those that are
-# redirected to the `figure`-call.
-SUBPLOTS_OWN_KWARGS = set(inspect.signature(plt.subplots).parameters.keys())
-SUBPLOTS_OWN_KWARGS.remove('fig_kw')
-FIGURE_KWARGS = set(inspect.signature(plt.figure).parameters.keys())
-FIGURE_KWARGS.remove('kwargs')
-SUBPLOTS_KWARGS = SUBPLOTS_OWN_KWARGS.union(FIGURE_KWARGS)
+if TYPE_CHECKING:
+    AxesTuple = Tuple["Axes", "Colorbar"]
+    AxesTupleList = Tuple[List["Axes"], List[Optional["Colorbar"]]]
```
Comment (jenshnielsen) [inline qcodes/dataset/plotting.py:46]: Since these are quoted we can get rid of the if typeckecking logic

**Comment by edumur** [comment]
Done.

I did not get what you meant by "documentation about where to import it from"...

File: qcodes/__init__.py (hunk 6 above)
Comment (jenshnielsen) [reply qcodes/__init__.py:56]: @edumur In the future please comment directly on the thing that is unclear. That makes the conversation easier to follow. I mean in current qcodes the user can do `from qcodes import test_instrument` but with this change they will have to do 
`from qcodes.instrument_drivers.test import test_instrument` etc. 

We should document this

Comment (edumur) [reply qcodes/__init__.py:56]: Sorry for that!

My issue was more like where and how do you want me to do that.
I looked in the doc and currently there is not so much about theses functions: https://qcodes.github.io/Qcodes/search.html?q=test_instrument&check_keywords=yes&area=default.

Thanks!

Comment (jenshnielsen) [reply qcodes/__init__.py:56]: I just meant that we should have a breaking change note like the one that you added now. 

```
From now `test_intrument` and `test_intruments` must be imported from
`qcodes.instrument_drivers.test`.
```

I would perhaps just make it a bit more explicit. Like

```
The `test_intrument` and `test_intruments` functions are no longer available from the toplevel qcodes namespace. 
If you require these functions they must be imported from `qcodes.instrument_drivers.test`. 
```

or something like that

**Comment by jenshnielsen** [comment]
bors merge

**Review by jenshnielsen — APPROVED** [review]

**Comment by jenshnielsen** [comment]
Looks good. thanks @edumur

**Comment by jenshnielsen** [comment]
bors merge

**Comment by jenshnielsen** [comment]
bors merge

**Pull request merged**